File: src/main/java/Q395_LongestSubstringwithAtLeastKRepeatingCharacters.java
import java.util.Arrays;

/***
 * 34mins medium
 * Runtime: 1 ms, faster than 83.76% of Java online submissions for Longest Substring with At Least K Repeating Characters.
 * Memory Usage: 39.4 MB, less than 26.25% of Java online submissions for Longest Substring with At Least K Repeating Characters.
 * can be further optimized by using char array and index instead of copying strings
 *
 */
public class Q395_LongestSubstringwithAtLeastKRepeatingCharacters {
    int k;
    int result = 0;
    public int longestSubstring(String s, int k) {
        this.k = k;
        rec(s);
        return result;
    }

    public void rec(String s) {
        if(s == null || s.length() == 0) {
            return;
        }
        int[] bs = new int[26];
        Arrays.fill(bs, 0);
        char[] c = s.toCharArray();
        for(int i = 0; i < c.length; i++) {
            bs[c[i] - 'a']++;
        }
        boolean flag = true;
        for(int i = 0; i < 26; i++) {
            if(bs[i] != 0 && bs[i] < k) {
                flag = false;
            }
        }
        if(flag) {
            result = Math.max(result, s.length());
            return;
        }
        StringBuilder sb = new StringBuilder();
        for(int i = 0; i < c.length; i++) {
            if(bs[c[i] - 'a'] < k) {
                rec(sb.toString());
                sb.setLength(0);
            }
            else {
                sb.append(c[i]);
            }
        }
        if(sb.length() > 0) {
            rec(sb.toString());
        }
    }
}

File: LeetCodePractice/LeetCodePractice/Q395_longestSubstringWithAtLeastKRepeatingCharacters.cpp
#include "LCHeader.h"
//11:32 11:55, bugs


class Solution {//11:32
public:
    int longestSubstring(string s, int k) {
        if (k == 1) {
            return s.size();
        }
        if (s == "") {
            return 0;
        }
        int n = s.size();
        vector<vector<int>> cum(n, vector<int>(26, 0));
        vector<int> pos(26, 0);
        cum[0][s[0] - 'a']++;
        int result = 0;
        for (int i = 1; i < n; i++) {
            cum[i] = cum[i - 1];
            cum[i][s[i] - 'a']++;
            int latest = -1;
            pos[s[i] - 'a'] = i;
            for (int j = 0; j < 26; j++) {
                if (cum[i][j] == 0 || cum[i][j] >= k) {
                    continue;
                }
                else {
                    latest = std::max(latest, pos[j]);
                }
            }
            if (latest != -1) {
                bool can = true;
                for (int j = 0; j < 26; j++) {
                    if (cum[i][j] - cum[latest][j] == 0) {
                        continue;
                    }
                    if (cum[i][j] - cum[latest][j] < k) {
                        can = false;
                        break;
                    }
                }
                if (can) {
                    result = std::max(result, i - latest);
                }
            }
            else {
                result = std::max(result, i - latest);
            }
        }
        return result;
    }
};